crates/noirc_frontend/src/parser/errors.rs

```diff
@@ -1,98 +1,152 @@
+            None => {
+                let mut expected = self
+                    .expected_tokens
+                    .iter()
+                    .map(ToString::to_string)
+                    .collect::<Vec<_>>();
+                expected.append(&mut self.expected_labels.iter().cloned().collect());
+
+                let primary = if expected.is_empty() {
+                    format!("Unexpected {} in input", self.found)
+                } else if expected.len() == 1 {
+                    format!(
+                        "Expected a {} but found {}",
+                        expected.first().unwrap(),
+                        self.found
+                    )
+                } else {
+                    let expected = expected
+                        .iter()
+                        .map(ToString::to_string)
+                        .collect::<Vec<_>>()
+                        .join(", ");
+
+                    format!("Unexpected {}, expected one of {}", self.found, expected)
                 };
 
-                let mut diag = Diagnostic::simple_error(
-                    "path is ambiguous".to_string(),
-                    "path contains a single keyword".to_string(),
-                    *span,
-                );
-                diag.add_note(note.to_owned());
-                diag
+                Diagnostic::simple_error(primary, String::new(), self.span)
             }
         }
     }
 }
+
+impl From<LexerErrorKind> for ParserError {
+    fn from(error: LexerErrorKind) -> Self {
+        let span = error.span();
+        ParserError {
+            expected_tokens: BTreeSet::new(),
+            expected_labels: BTreeSet::new(),
+            found: Token::EOF,
+            lexer_errors: vec![error],
+            reason: None,
+            span,
+        }
+    }
+}
+
+impl chumsky::Error<Token> for ParserError {
+    type Span = Span;
+    type Label = String;
+
+    fn expected_input_found<Iter>(span: Self::Span, expected: Iter, found: Option<Token>) -> Self
+    where
+        Iter: IntoIterator<Item = Option<Token>>,
+    {
+        ParserError {
+            expected_tokens: expected
+                .into_iter()
+                .map(|opt| opt.unwrap_or(Token::EOF))
+                .collect(),
+            expected_labels: BTreeSet::new(),
+            found: found.unwrap_or(Token::EOF),
+            lexer_errors: vec![],
+            reason: None,
+            span,
+        }
+    }
+
+    fn with_label(mut self, label: Self::Label) -> Self {
+        self.expected_labels.insert(label);
+        self
+    }
+
+    fn merge(mut self, mut other: Self) -> Self {
+        self.expected_tokens.append(&mut other.expected_tokens);
+        self.expected_labels.append(&mut other.expected_labels);
+        self.lexer_errors.append(&mut other.lexer_errors);
+
+        if self.reason.is_none() {
+            self.reason = other.reason;
+        }
+
+        // Is this needed?
```

Comment:
Can this be resolved? I think we only ever need to merge spans for Tokens/Expressions/Statements and not statements